NET-1825: New ACL token creation docs (#16465)
hashicorp/consul: website/content/docs/security/acl/tokens/create/create-a-service-token.mdx

---
layout: docs
page_title: Create tokens for service registration
description: >-
  Learn how to create ACL tokens that your services can present to Consul servers so that they can register with the Consul catalog.
---

# Create a service token

> This topic describes how to create a token that you can use to discover services and to register services into the Consul catalog. If you are using Consul service mesh, this token can be used by a sidecar proxy to discover and route to other services.

Author: I think I fixed this wording:

> This topic describes how to create a token that you can use to register a service and discover services in the Consul catalog. If you are using Consul service mesh, this token can be used by a sidecar proxy to discover and route to other services.

I opted for "and discover services in the Catalog" because that is actually the permission it grants (i.e. it is not limited to just the dependencies of that service, although that is the purpose of those permissions)

## Introduction

Services must present a token linked to policies that grant the appropriate set of permissions. For a service to be discoverable or to interact with other services in a mesh, the token must be linked to policies that grant the following permissions:

> * `service:write`: Enables the service to update the catalog. If service mesh is enabled, the service's sidecar proxy can also update the catalog.Note that this permission implicitly grants `intention:read` permission to sidecar proxies so that they can read and enforce intentions. Refer to [Intention Management Permissions](/consul/docs/connect/intentions#intention-management-permissions) for details.

Reviewer: ```suggestion
* `service:write`: Enables the service to update the catalog. If service mesh is enabled, the service's sidecar proxy can also update the catalog. Note that this permission implicitly grants `intention:read` permission to sidecar proxies so that they can read and enforce intentions. Refer to [Intention Management Permissions](/consul/docs/connect/intentions#intention-management-permissions) for details.
```

The following example creates a token linked to a service identity named `svc1`:

```shell-session
$ curl --request PUT http://127.0.0.1:8500/v1/acl/token \
  --header "X-Consul-Token: $CONSUL_HTTP_TOKEN" \
  --data '{
  "ServiceIdentities": [
    {
      "ServiceName": "svc1"
    }
  ]
}'
```

</Tab>

</Tabs>

## Service identity in Consul Enterprise <EnterpriseAlert inline />

Refer to [Service identities](/consul/docs/security/acl#service-identities) for information about creating service identities that you can link to tokens.

You can manually create tokens using the Consul command line or API endpoint. You can also enable Consul to dynamically create tokens from trusted external systems using an auth method.

<Tabs>

<Tab heading="CLI" group="CLI">

Run the `consul acl token create` command and specify the policy or service identity to link to create a token. Refer to [Consul ACL Token Create](/consul/commands/acl/token/create) for details about the `consul acl token create` command.

> You can specify an admin partition, namespace, or both when creating tokens in Consul Enterprise. The token is only valid in the specified network areas. The following example creates an ACL token that the service can use to register in the `ns1` namespace of partition `ptn1`:

Reviewer: "Network areas" are a specific enterprise feature that do not apply here. We should not use that term. I suggest "scopes" instead.

```suggestion
You can specify an admin partition, namespace, or both when creating tokens in Consul Enterprise. The token can only include permissions in the specified scope, if any. The following example creates an ACL token that the service can use to register in the `ns1` namespace of partition `ptn1`:
```

Author: I went through and replaced "network areas" with "scopes".

You can specify an admin partition, namespace, or both when creating tokens in Consul Enterprise. The token is only valid in the specified network areas. The following example creates an ACL token that the service can use to register in the `ns1` namespace of partition `ptn1`:

```shell-session
$ curl --request PUT http://127.0.0.1:8500/v1/acl/token \
  --header "X-Consul-Token: $CONSUL_HTTP_TOKEN" \
  --data '{
  "ServiceIdentities": [
    {
      "ServiceName": "svc1"
    }
  ],
  "Namespace": "ns1",
  "Partition": "ptn1"
}'
```

</Tab>

</Tabs>

## Custom policy in Consul OSS

When you are unable to link tokens to a service identity, you can define policies, register them with Consul, and link the policies to tokens that enable services to register into the Consul catalog.

### Define a policy

You can send policy definitions as command line or API arguments or define them in an external HCL or JSON file. The following example policy is defined in a file. Refer to [ACL Rules](/consul/docs/security/acl/acl-rules) for details about all of the rules you can use in your policies.

> The following example policy grants the `svc1` service `write` permissions so that it can register into the catalog. For service mesh, the policy grants the `svc1-sidecar-proxy` service `write` permissions so that the sidecar proxy can register into the catalog. It grants service and node `read` permissions to discover and route to other services.

Author: I did leave `Refer to [ACL Rules](/consul/docs/security/acl/acl-rules) ...` in the first paragraph though.

$ consul acl policy create \
    -name "svc1-register" -rules @svc1-register.hcl \
    -description "Allow svc1 to register into the catalog"
```

Refer to [Consul ACL Policy Create](/consul/commands/acl/policy/create) for details about the `consul acl token create` command.

</Tab>

<Tab heading="API" group="API">

Send a PUT request to the `/acl/policy` endpoint and specify the policy rules in the request body to create a policy. The following example registers the policy defined in `svc1-register.hcl`. You must embed policy rules in the `Rules` field of the request body

```shell-session
$ curl --request PUT http://127.0.0.1:8500/v1/acl/policy \
  --header "X-Consul-Token: $CONSUL_HTTP_TOKEN" \
  --data '{
  "Name": "svc1-register",
  "Description": "Allow svc1 to register into the catalog",
  "Rules": "service \"svc1\" {\n  policy = \"write\"\n}\nservice \"svc1-sidecar-proxy\" {\n  policy = \"write\"\n}\nservice_prefix \"\" {\n  policy = \"read\"\n}\nnode_prefix \"\" {\n  policy = \"read\"\n}\n"
}'
```

Refer to [ACL Policy HTTP API](/consul/api-docs/acl/policies) for additional information about using the API endpoint.

</Tab>

</Tabs>

### Link the policy to a token

> After registering the policies into Consul, you can create and link tokens using the Consul command line or API endpoint. You can also enable Consul to dynamically create tokens from trusted external systems using an auth method.

Reviewer: Should we link to somewhere in the auth method docs?
